### PR TITLE
[dif/csrng] Finish implementing missing funtions in dif_csrng

### PR DESCRIPTION
**Pull request by mcy**

### CODE REVIEW

File: sw/device/lib/dif/dif_csrng.h [hunk 1]
```diff
@@ -319,6 +429,61 @@ OT_WARN_UNUSED_RESULT
...
+ */
+OT_WARN_UNUSED_RESULT
+dif_result_t dif_csrng_get_cmd_force_error(const dif_csrng_t *csrng,
+                                           dif_csrng_error_t error);
+
+/**
+ * Returns an opaque blob indicating the main state machine's current state.
+ *
+ * @param csrng An CSRNG handle
+ * @param state The state machine state as an opaque blob.
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_csrng.h:459]: nit:
```suggestion
 * @param[out] state The state machine state as an opaque blob.
```

File: sw/device/lib/dif/dif_csrng.h [hunk 2]
```diff
@@ -319,6 +429,61 @@ OT_WARN_UNUSED_RESULT
...
+OT_WARN_UNUSED_RESULT
+dif_result_t dif_csrng_get_main_state_machine(const dif_csrng_t *csrng,
+                                              uint32_t *state);
+
+/**
+ * Returns a bitset indicating which hardware CSRNGs have encountered
+ * exceptions.
+ *
+ * @param csrng An CSRNG handle
+ * @param exceptions The bitset of exception states.
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_csrng.h:471]:
```suggestion
 * @param[out] exceptions The bitset of exception states.
```

File: sw/device/lib/dif/dif_csrng.c [hunk 3]
```diff
@@ -230,16 +222,211 @@ dif_result_t dif_csrng_get_cmd_interface_status(
...
+dif_result_t dif_csrng_get_hw_csrng_exceptions(const dif_csrng_t *csrng,
+                                               uint32_t *exceptions) {
+  if (csrng == NULL || exceptions == NULL) {
+    return kDifBadArg;
+  }
+
+  *exceptions =
+      mmio_region_read32(csrng->base_addr, CSRNG_HW_EXC_STS_REG_OFFSET);
+  return kDifOk;
+}
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_csrng.c:423]: Nit. Add a new line here for consistency.

File: sw/device/lib/dif/dif_csrng.h [hunk 4]
```diff
@@ -263,17 +373,17 @@ dif_result_t dif_csrng_update(const dif_csrng_t *csrng,
...
- * This function blocks until all data is read from the CSRNG. The caller is
- * responsible for making sure that the `len` parameter matches the length set
- * in the `dif_csrng_generate_start()` call.
+ * This function reads `len` words out of the CSRNG. This function should be
+ * called repeatedly until the number of words requested in
+ * `dif_csrng_generate_start()` is exhausted. This function will block until
+ * `len` words are read.
  *
  * `dif_csrng_get_output_status()` can be called before this function to ensure
  * there is data availble in the CSRNG output buffer.
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_csrng.h:389]: Would you mind fixing this typo?
```suggestion
 * there is data available in the CSRNG output buffer.
```
 * there is data availble in the CSRNG output buffer.

File: sw/device/lib/dif/dif_csrng.h [hunk 5]
```diff
@@ -319,6 +429,61 @@ OT_WARN_UNUSED_RESULT
...
+ * @param csrng An CSRNG handle
+ * @param exceptions[out] The bitset of exception states.
+ * @return The result of the operation.
+ */
+OT_WARN_UNUSED_RESULT
+dif_result_t dif_csrng_get_hw_csrng_exceptions(const dif_csrng_t *csrng,
+                                               uint32_t *exceptions);
+
+/**
+ * Clears recorded hardware CSRNG exceptions
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_csrng.h:479]: Nit. Missing full stop.

File: sw/device/lib/dif/dif_csrng.h [hunk 6]
```diff
@@ -378,6 +542,29 @@ dif_result_t dif_csrng_is_locked(const dif_csrng_t *csrng, bool *is_locked);
 OT_WARN_UNUSED_RESULT
 dif_result_t dif_csrng_stop(const dif_csrng_t *csrng);
 
+/**
+ * Gets the recoverable alerts currently recorded in the CSRNG block
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_csrng.h:546]: Nit. Missing full stop here.

File: sw/device/lib/dif/dif_csrng.h [hunk 7]
```diff
@@ -378,6 +542,29 @@ dif_result_t dif_csrng_is_locked(const dif_csrng_t *csrng, bool *is_locked);
...
+ * @param csrng A CSRNG handle.
+ * @param[out] alerts Bitset of alerts currently recorded.
+ * @return The result of the operation.
+ */
+OT_WARN_UNUSED_RESULT
+dif_result_t dif_csrng_get_recoverable_alerts(const dif_csrng_t *csrng,
+                                              uint32_t *alerts);
+
+/**
+ * Clears all recoverable alerts currently recorded in the CSRNG block
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_csrng.h:560]: Nit. Missing full stop here.

**Review by timothytrippel — APPROVED** [review]

**Review by engdoreis — APPROVED** [review]

**Pull request merged**